### PR TITLE
Automatically move differentials to the left side

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Fixes https://github.com/SciML/ModelingToolkit.jl/issues/860

### CODE REVIEW

File: src/systems/diffeqs/odesystem.jl [hunk 1]
```diff
@@ -92,6 +92,31 @@ struct ODESystem <: AbstractODESystem
...
 end
 
+function move_diffs(eq::Equation,r)
+    # Do not modify `D(x) ~ ...`, already correct
+    # Ignore `δ(x) ~ ...` for now
+    if !(eq.lhs isa Term && ModelingToolkit.operation(eq.lhs) isa Differential) &&
+       !(eq.lhs isa Term && ModelingToolkit.operation(eq.lhs) isa Difference)
+       _eq = eq.rhs-eq.lhs
+       rhs = r(_eq)
+       if isnothing(rhs)
```
Comment (YingboMa) [inline src/systems/diffeqs/odesystem.jl:102]:
```suggestion
       if rhs === nothing
```
`isnothing` probably should never be used, and just removed from Base.

File: src/systems/diffeqs/odesystem.jl [hunk 2]
```diff
@@ -92,6 +92,31 @@ struct ODESystem <: AbstractODESystem
     end
 end
 
+function move_diffs(eq::Equation,r)
+    # Do not modify `D(x) ~ ...`, already correct
+    # Ignore `δ(x) ~ ...` for now
+    if !(eq.lhs isa Term && ModelingToolkit.operation(eq.lhs) isa Differential) &&
```
Comment (YingboMa) [inline src/systems/diffeqs/odesystem.jl:98]: This code already lives in MTK.

File: test/odesystem.jl [hunk 3]
```diff
@@ -439,10 +439,39 @@ function periodic_difference_affect!(int)
...
+D = Differential(t)
+eqs = [0 ~ x + y - D(x)
+       x ~ y]
+sys = ODESystem(eqs)
+eqs2 = equations(sys)
+eqs3 = [D(x) ~ x + y,
+        0 ~ y-x]
+@test isequal(eqs2,eqs3)
+
+eqs = [0 ~ x + y + D(x)
```
Comment (YingboMa) [inline test/odesystem.jl:462]: What happens if there's `D(x)/x ~ x`?

Comment (ChrisRackauckas) [reply test/odesystem.jl:462]: `D(x)/x ~ x` will be untouched like before. `D(x)*x ~ x` will give you `-D(x)*x ~ -x`, which is a little odd but works fine for `DAEProblem`.

Comment (YingboMa) [reply test/odesystem.jl:462]: Could you test these cases as well? Also `sin(D(x) * x) ~ x` etc.

**Comment by ChrisRackauckas** [comment]
@YingboMa one last test now throws the wrong error. It errored before, but it shouldn't throw this error.

```julia
julia> equations(sys)
3-element Vector{Equation}:
 0 ~ x(t) + sin(u1(t) + u2(t))
 Differential(t)(x(t)) ~ x(t) + y1(t)
 0 ~ sin(y2(t)) - cos(x(t))
fullvars = Any[Differential(t)(x(t)), x(t), u1(t), u2(t), y2(t)]
```

How come `y2(t)` was changed from a parameter to a state in structural_simplify? That seems like it should never be able to happen?

**Comment by YingboMa** [comment]
Maybe rule drops the metadata.

**Comment by YingboMa** [comment]
What's the MWE?

**Comment by ChrisRackauckas** [comment]
```julia
# Example 5 from Pantelides' original paper
@parameters t
params = collect(@parameters y1(t) y2(t))
sts = collect(@variables x(t) u1(t) u2(t))
D = Differential(t)
eqs = [
       0 ~ x + sin(u1 + u2)
       D(x) ~ x + y1
       cos(x) ~ sin(y2)
      ]
@named sys = ODESystem(eqs, t, sts, params)
@test_throws ModelingToolkit.InvalidSystemException structural_simplify(sys)
```

**Pull request closed without merging**